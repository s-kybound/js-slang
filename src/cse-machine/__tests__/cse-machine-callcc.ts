import { Chapter, Variant } from '../../types'
import { expectResult } from '../../utils/testing'

// as continuations mostly target the scheme implementation, we will test continuations
// using a scheme context.
const optionECScm = { chapter: Chapter.FULL_SCHEME, variant: Variant.EXPLICIT_CONTROL }

test('basic call/cc works', () => {
  return expectResult(
    `
    (+ 1 2 (call/cc (lambda (k) (k 3))) 4)
  `,
    optionECScm
  ).toMatchInlineSnapshot(`10`)
})

test('call/cc can be used to escape a computation', () => {
  return expectResult(
    `
    (define test 1)
    (call/cc (lambda (k) 
                (set! test 2) 
                (k 'escaped)
                (set! test 3)))
    ;; test should be 2
    test
  `,
    optionECScm
  ).toMatchInlineSnapshot(`2`)
})

test('call/cc can be stored as a value', () => {
  return expectResult(
    `
    ;; storing a continuation and calling it
    (define a #f)

    (+ 1 2 3 (call/cc (lambda (k) (set! a k))) 4 5)

    ;; as continuations are represented with dummy 
    ;; identity functions, we should not expect to see 6
    (a 6)
    `,
    optionECScm
  ).toMatchInlineSnapshot(`21`)
})

// test('even when stored as a value, calling a continuation should alter the execution flow', () => {
//   return expectResult(
//     `
//     ;; storing a continuation and calling it
//     (define a #f)

//     (+ 1 2 3 (call/cc (lambda (k) (set! a k))) 4 5)

//     ;; the following addition should be ignored

//     (+ (a 6) 7)
//     `,
//     optionECScm
//   ).toMatchInlineSnapshot(`21`)
// })