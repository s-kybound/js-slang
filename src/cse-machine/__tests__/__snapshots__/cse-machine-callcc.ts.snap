// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic call/cc works: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (+ 1 2 (call/cc (lambda (k) (k 3))) 4)
  ",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 10,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`call/cc can be stored as a value: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    ;; storing a continuation
    (define a #f)

    (+ 1 2 3 (call/cc (lambda (k) (set! a k) 0)) 4 5)

    ;; continuations are treated as functions
    ;; so we can do this:
    (procedure? a)
    ",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`call/cc can be used to escape a computation: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (define test 1)
    (call/cc (lambda (k) 
                (set! test 2) 
                (k 'escaped)
                (set! test 3)))
    ;; test should be 2
    test
  ",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;
